geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/data/AccumuloRasterQueryPlanner.scala

```diff
@@ -28,18 +28,35 @@ import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
 import org.opengis.feature.simple.SimpleFeatureType
 import org.opengis.filter.Filter
 
+import scala.annotation.tailrec
 import scala.collection.JavaConversions._
 import scala.util.Try
 
-// TODO: Consider adding resolutions + extent info  https://geomesa.atlassian.net/browse/GEOMESA-645
-class AccumuloRasterQueryPlanner extends Logging with IndexFilterHelpers {
-  import AccumuloRasterQueryPlanner._
+object AccumuloRasterQueryPlanner extends Logging with IndexFilterHelpers {
 
-  def getQueryPlan(rq: RasterQuery, resAndGeoHashMap: ImmutableSetMultimap[Double, Int]): Option[QueryPlan] = {
-    val availableResolutions = resAndGeoHashMap.keySet().toList.sorted
+  def improvedOverlaps(a: Envelope, b: Envelope): Boolean = a.intersects(b) && a.intersection(b).getArea != 0.0
 
-    // Step 1. Pick resolution
-    val selectedRes: Double = selectResolution(rq.resolution, availableResolutions)
+  def getAcceptableResolution(rq: RasterQuery, resAndBoundsMap: ImmutableMap[Double, BoundingBox]): Option[Double] = {
+    val availableResolutions = resAndBoundsMap.keySet().toList.sorted
+    val preferredRes: Double = selectResolution(rq.resolution, availableResolutions)
+    getCourserBounds(rq.bbox, preferredRes, resAndBoundsMap)
+  }
+
+  @tailrec
+  def getCourserBounds(qb: BoundingBox, r: Double, rToB: ImmutableMap[Double, BoundingBox]): Option[Double] = {
+    // get finest bounds that satisfies our needs
+    if (rToB.containsKey(r)) {
+      if (improvedOverlaps(qb.envelope, rToB.get(r).envelope)) Some(r) else {
+        val cr = rToB.keys.filter(_ > r)
+        if (cr.nonEmpty) getCourserBounds(qb, cr.min, rToB) else None
+      }
+    } else None
+  }
```

Comment:
Rather than having this function be tail recursive, does something like this work and read more plainly?
```
rToB.keys.filter(_ >= r).sort.reverse.find{ range => improvedOverlaps(qb.envelope, rToB(range).envelope) }
} else None